p2pserver/message/types/address.go

```diff
@@ -32,44 +33,70 @@ type Addr struct {
 }
 
 //Serialize message payload
-func (this Addr) Serialization() ([]byte, error) {
-	p := new(bytes.Buffer)
+func (this Addr) Serialization(sink *common.ZeroCopySink) error {
 	num := uint64(len(this.NodeAddrs))
-	err := binary.Write(p, binary.LittleEndian, num)
-	if err != nil {
-		return nil, errors.NewDetailErr(err, errors.ErrNetPackFail, fmt.Sprintf("write error. num:%v", num))
-	}
+	sink.WriteUint64(num)
 
-	err = binary.Write(p, binary.LittleEndian, this.NodeAddrs)
-	if err != nil {
-		return nil, errors.NewDetailErr(err, errors.ErrNetPackFail, fmt.Sprintf("write error. NodeAddrs:%v", this.NodeAddrs))
+	for _, addr := range this.NodeAddrs {
+		sink.WriteInt64(addr.Time)
+		sink.WriteUint64(addr.Services)
+		sink.WriteBytes(addr.IpAddr[:])
+		sink.WriteUint16(addr.Port)
+		sink.WriteUint64(addr.ID)
 	}
 
-	return p.Bytes(), nil
+	return nil
 }
 
 func (this *Addr) CmdType() string {
 	return comm.ADDR_TYPE
 }
 
-//Deserialize message payload
-func (this *Addr) Deserialization(p []byte) error {
-	buf := bytes.NewBuffer(p)
+func (this *Addr) Deserialization(source *common.ZeroCopySource) error {
+	count, eof := source.NextUint64()
+	if eof {
+		return io.ErrUnexpectedEOF
+	}
+
+	for i := 0; i < int(count); i++ {
+		var addr comm.PeerAddr
+		addr.Time, eof = source.NextInt64()
+		addr.Services, eof = source.NextUint64()
+		buf, _ := source.NextBytes(uint64(len(addr.IpAddr[:])))
+		copy(addr.IpAddr[:], buf)
+		addr.Port, eof = source.NextUint16()
+		addr.ID, eof = source.NextUint64()
+		if eof {
+			return io.ErrUnexpectedEOF
+		}
 
+		this.NodeAddrs = append(this.NodeAddrs, addr)
+	}
+
+	if count > comm.MAX_ADDR_NODE_CNT {
```

Comment:
1.  the param is not `[]byte` any more, so all bytes belong to this message should be read out from the `source`, to avoid subtle bug. 2. the max payload size has checked, so this safe.

Reply:
so this is safe now, should we remove this count limit??

Reply:
since there is a global max outbound connection limit, the MAX_ADDR_NODE_CNT is to avoid  malicious node send too many addresses and occupy the outbound connection?